File: apps/renderdata.cpp
/*
 * This file is part of Co-Section.
 *
 * Copyright (C) 2020 Embodied Vision Group, Max Planck Institute for Intelligent Systems, Germany.
 * Developed by Michael Strecke <mstrecke at tue dot mpg dot de>.
 * For more information see <https://cosection.is.tue.mpg.de/>.
 * If you use this code, please cite the respective publication as
 * listed on the website.
 *
 * Co-Section is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Co-Section is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Co-Section.  If not, see <https://www.gnu.org/licenses/>.
 */
#include <iostream>
#include <fstream>

#include <sstream>
#include <iomanip>

#include <opencv2/opencv.hpp>
#ifdef HAVE_OPENCV_VIZ
#include <opencv2/viz.hpp>
#endif

#include <sophus/se3.hpp>
#include <opencv2/core/eigen.hpp>

#include <boost/program_options.hpp>

namespace po = boost::program_options;

cv::viz::Viz3d window;
cv::Mat colorMap;

int frameIncr = 1;

void KeyboardCallback ( const cv::viz::KeyboardEvent &w, void *t ) {
    cv::viz::Viz3d *win = ( cv::viz::Viz3d* ) t;

    if ( w.action ) {
        if ( w.symbol == "Left" ) {
            frameIncr = -1;
        }
        if ( w.symbol == "Right" ) {
            frameIncr = 1;
        }
        if ( w.code == ' ' ) {
            frameIncr = 0;
        }
        if ( w.code == 'I' || w.code == 'i' ) {
            cv::Affine3f viewPose = win->getViewerPose();
            std::cout << "\nViewer pose:\n" << viewPose.matrix << std::endl;
        }
    }
}

cv::Mat randomColors () {
    cv::Mat hsv ( 1, 256, CV_32FC3 );

    for ( int i = 1; i < 256; ++i ) {
        hsv.at<cv::Vec3f> ( i ) = cv::Vec3f ( ( i / 256.f ) * 360.f, 1.f, 1.f );
    }

    cv::Mat rgb;
    cv::cvtColor ( hsv, rgb, cv::COLOR_HSV2RGB );

    rgb.convertTo ( rgb, CV_8U, 255 );

    cv::RNG rng ( 6893 );
    randShuffle ( rgb, 1, &rng );

    rgb.at<cv::Vec3b> ( 0 ) = cv::Vec3b ( 255, 255, 255 );

    return rgb;
}

void show3DVis ( const cv::Affine3f& campose, const cv::Matx33f& intr,
                 const cv::Mat& rendered, const cv::viz::Mesh& bg_mesh,
                 const std::map<int, cv::Affine3f >& poses,
                 const std::map<int, cv::viz::Mesh >& meshes,
                 const bool followcampose ) {
    window.removeAllWidgets();
    window.removeAllLights();
    window.addLight ( cv::Vec3d ( 0,-10,-10 ) );

    if ( followcampose )
        window.setViewerPose ( campose );
    else {
        cv::viz::WCameraPosition camPos ( intr, rendered, 0.1 );
        camPos.applyTransform ( campose );
        window.showWidget ( "CamPos", camPos );
    }
    for ( const auto& mesh : meshes ) {
        cv::Affine3f obj_pose = poses.at ( mesh.first );
        cv::viz::Mesh m_obj = mesh.second;
        if ( m_obj.cloud.rows == 1 ) {
            cv::viz::WMesh m_show_obj ( m_obj );
            m_show_obj.setColor (
                cv::viz::Color ( colorMap.at<cv::Vec3b> ( 0, mesh.first ) ) );
            m_show_obj.applyTransform ( obj_pose );
            std::stringstream mesh_str;
            mesh_str << "Mesh " << mesh.first;
            window.showWidget ( mesh_str.str(), m_show_obj );
            window.setRenderingProperty ( mesh_str.str(), cv::viz::SHADING,
                                          cv::viz::SHADING_PHONG );
        }
    }

    if ( bg_mesh.cloud.rows == 1 ) {
        cv::viz::WMesh viz_mesh ( bg_mesh );
        viz_mesh.applyTransform (
            cv::Affine3f().translate ( cv::Vec3f ( 0, 0, 2.56f ) ) );
        window.showWidget ( "mesh", viz_mesh );
        window.setRenderingProperty ( "mesh", cv::viz::SHADING,
                                      cv::viz::SHADING_PHONG );
    }

    window.spinOnce();
}

void main_loop ( const std::string& dirpath = "", const std::string& opath = "",
                 const int startframe = 0, const int endframe = -1,
                 const bool tsdf_flag = false,
                 const bool followcampose = true ) {
    cv::Size frameSize = cv::Size ( 640, 480 );
    float fx, fy, cx, cy;
    fx = fy = 525.f;
    cx = frameSize.width/2 - .5f;
    cy = frameSize.height/2 - .5f;
    cv::Matx33f intr = cv::Matx33f ( fx,  0, cx,
                                     0, fy, cy,
                                     0,  0,  1 );
    std::ifstream calibstr ( dirpath + "/calibration.txt" );
    if ( calibstr.is_open() ) {
        calibstr >> intr ( 0, 0 )
                 >> intr ( 1, 1 )
                 >> intr ( 0, 2 )
                 >> intr ( 1, 2 );

        calibstr >> frameSize.width
                 >> frameSize.height;
    }
    calibstr.close();
    cv::viz::Camera cam ( intr, frameSize );
    window.setCamera ( cam );

    if ( !opath.empty() )
        window.setOffScreenRendering();

    // read in poses
    std::map<int, cv::Affine3f> campose;

    int lastframe = -1;
    std::ifstream infile ( dirpath + std::string ( "/poses-cam.txt" ) );
    if ( infile.is_open() ) {
        std::string line;
        while ( std::getline ( infile, line ) ) {
            std::istringstream iss ( line );
            int t;
            float tx, ty, tz, qx, qy, qz, qw;
            if ( ! ( iss >> t >> tx >> ty >> tz >> qx >> qy >> qz >> qw ) ) {
                break;
            }

            Sophus::SE3f se3_pose;
            se3_pose.setQuaternion ( Eigen::Quaternionf ( qw, qx, qy, qz ) );
            se3_pose.translation() = Eigen::Vector3f ( tx, ty, tz );

            cv::Affine3f::Mat4 affine;
            cv::eigen2cv ( se3_pose.matrix(), affine );

            campose[t] = cv::Affine3f ( affine );
            lastframe = t;
        }
    }

    std::map<int, std::map<int, cv::Affine3f >> poses;
    size_t obj = 1;
    while ( true ) {
        std::ifstream infile ( dirpath + std::string ( "/poses-" )
                               + std::to_string ( obj )
                               + std::string ( ".txt" ) );
        if ( infile.is_open() ) {
            poses[obj] = std::map<int, cv::Affine3f>();
            std::string line;
            while ( std::getline ( infile, line ) ) {
                std::istringstream iss ( line );
                int t;
                float tx, ty, tz, qx, qy, qz, qw;
                if ( ! ( iss >> t >> tx >> ty >> tz >> qx >> qy >> qz >> qw ) )
                    break;

                Sophus::SE3f se3_pose;
                se3_pose.setQuaternion (
                    Eigen::Quaternionf ( qw, qx, qy, qz ) );
                se3_pose.translation() = Eigen::Vector3f ( tx, ty, tz );

                cv::Affine3f::Mat4 affine;
                cv::eigen2cv ( se3_pose.matrix(), affine );

                poses[obj][t] = cv::Affine3f ( affine );
            }
        } else
            break;

        obj++;
    }

    int frame = startframe, last_keyframe = -1;
    if ( startframe == endframe && opath.empty() )
        frameIncr = 0;
    cv::viz::Mesh bg_mesh;
    std::map<int, cv::viz::Mesh> meshes;
    do {
        int newFrame = tsdf_flag ? frame : ( frame / 10 ) * 10;
        if ( newFrame != last_keyframe ) {
            last_keyframe = newFrame;
            // read in meshes
            std::stringstream bg_meshfile;
            if ( tsdf_flag )
                bg_meshfile << dirpath << "/frame_meshes/bg/" << std::setw ( 4 )
                            << std::setfill ( '0' ) << last_keyframe << ".ply";
            else
                bg_meshfile << dirpath << "/optim_meshes/bg/" << std::setw ( 4 )
                            << std::setfill ( '0' ) << last_keyframe << ".ply";
            bg_mesh = cv::viz::Mesh::load ( bg_meshfile.str() );
            meshes.clear();

            size_t obj = 1;
            for ( obj = 1; obj <= poses.size(); ++obj ) {
                std::stringstream meshfile;
                if ( tsdf_flag )
                    meshfile << dirpath << "/frame_meshes/" << obj << "/"
                             << std::setw ( 4 ) << std::setfill ( '0' )
                             << last_keyframe << ".ply";
                else
                    meshfile << dirpath << "/optim_meshes/" << obj << "/"
                             << std::setw ( 4 ) << std::setfill ( '0' )
                             << last_keyframe << ".ply";
                std::ifstream infile ( meshfile.str() );
                if ( infile.is_open() ) {
                    infile.close();

                    meshes[obj] = cv::viz::Mesh::load ( meshfile.str() );
                }
            }
        }

        std::stringstream renderfile;
        renderfile << dirpath << "/output/" << std::setw ( 4 )
                   << std::setfill ( '0' ) << frame << ".png";
        cv::Mat rendered = cv::imread ( renderfile.str() );

        std::map<int, cv::Affine3f> obj_poses;
        for ( const auto& mesh : meshes ) {
            if ( poses[mesh.first].count ( frame ) )
                obj_poses[mesh.first] = poses[mesh.first][frame];
        }
        show3DVis ( campose[frame], intr, rendered, bg_mesh, obj_poses, meshes,
                    followcampose );
        std::cout << "\rFrame: " << std::setw ( 4 ) << std::setfill ( '0' )
                  << frame << std::flush;
        if ( !opath.empty() ) {
            std::stringstream filename;
            filename << opath << "/" << std::setw ( 4 ) << std::setfill ( '0' )
                     << frame << ".png";
            if ( !imwrite ( filename.str(), window.getScreenshot() ) ) {
                std::cerr << "\nCould not write " << filename.str() << "!"
                          << " Does the output directory exist?" << std::endl;
                exit ( 1 );
            }
        }
        if ( frame + frameIncr <= lastframe && frame + frameIncr >= 0 )
            frame += frameIncr;
    } while ( ( frame <= endframe || opath.empty() ) && !window.wasStopped() );
    std::cout << std::endl;
}


int main ( int argc, char **argv ) {
    colorMap = randomColors();
    bool followcampose = false;

    po::options_description options ( "Render 3D models created by EM-Fusion or"
                                      " Co-Section" );

    po::options_description required ( "Required input" );
    required.add_options ()
    ( "dir,d", po::value<std::string>(),
      "EM-Fusion or Co-Section output directory" )
    ;

    po::options_description optional ( "Optional flags" );
    optional.add_options ()
    ( "help,h", "Print this help" )
    ( "opath,o", po::value<std::string>(),
      "Output path. Should be a folder. Files will be written as "
      "opath/<framenum>.png" )
    ( "frames,f", po::value<std::string>(),
      "Which frame(s) to render. Either single number n or range n-m" )
    ( "tsdf,t", "Render TSDF models instead of optimized ones" )
    ( "pose,p", po::value<std::string>(),
      "Pose from which to render the 3D models" )
    ( "outpose", po::value<std::string>(),
      "Output file to write final pose to (for reproducing results)" )
    ( "followcam", po::bool_switch ( &followcampose ),
      "Whether to follow the camera viewpoint of the recording" )
    ;

    options.add ( required ).add ( optional );
    po::variables_map result;
    po::store ( po::parse_command_line ( argc, argv, options ), result );
    po::notify ( result );

    if ( result.count ( "help" ) || ! ( result.count ( "dir" ) ) ) {
        std::cout << options << std::endl;
        exit ( 1 );
    }

    window = cv::viz::getWindowByName ( "3D Scene" );
    window.spinOnce();

    if ( result.count ( "dir" ) ) {
        std::string dirpath = result["dir"].as<std::string>();
        std::string opath;
        if ( result.count ( "opath" ) ) {
            opath = result["opath"].as<std::string>();
        } else {
            window.registerKeyboardCallback ( KeyboardCallback, &window );
        }

        int startframe = 0, endframe = -1;
        bool tsdf_flag = false;
        if ( result.count ( "frames" ) ) {
            std::istringstream framestr ( result["frames"].as<std::string>() );
            std::vector<std::string> frames;
            std::string f;
            while ( getline ( framestr, f, '-' ) )
                frames.push_back ( f );

            startframe = stoi ( frames[0] );
            if ( frames.size() > 1 )
                endframe = stoi ( frames[1] );
            else
                endframe = startframe;
        }
        if ( result.count ( "tsdf" ) )
            tsdf_flag = true;

        if ( ! followcampose ) {
            if ( result.count ( "pose" ) ) {
                std::ifstream posefile ( result["pose"].as<std::string>() );
                cv::Affine3d campose;
                for ( int i = 0; i < 16; ++i )
                    posefile >> campose.matrix.val[i];
                posefile.close();
                window.setViewerPose ( campose );
            } else {
                cv::Affine3d campose =
                    cv::Affine3d().translate ( cv::Vec3d ( 0, 0, -1 ) );
                window.setViewerPose ( campose );
            }
        }

        main_loop ( dirpath, opath, startframe, endframe, tsdf_flag,
                    followcampose );
    }

    cv::Affine3d v_pose = window.getViewerPose();
    std::cout << "Viewer pose at finish:" << std::endl;
    std::cout << v_pose.matrix << std::endl;
    if ( result.count ( "outpose" ) ) {
        std::ofstream posematfile ( result["outpose"].as<std::string>() );
        for ( int i = 0; i < 16; ++i ) {
            posematfile << v_pose.matrix.val[i];
            if ( ( i+1 ) % 4 == 0 )
                posematfile << "\n";
            else
                posematfile << " ";
        }
        posematfile.close();
    }
}

File: apps/renderdata.py
#
# This file is part of Co-Section.
#
# Copyright (C) 2020 Embodied Vision Group, Max Planck Institute for Intelligent Systems, Germany.
# Developed by Michael Strecke <mstrecke at tue dot mpg dot de>.
# For more information see <https://cosection.is.tue.mpg.de/>.
# If you use this code, please cite the respective publication as
# listed on the website.
#
# Co-Section is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Co-Section is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Co-Section.  If not, see <https://www.gnu.org/licenses/>.
#
import bpy
import bpy_extras
#import utils
import mathutils

import argparse
import math
import os
import sys
import random


colors = [[255, 255, 255],
          [255, 12, 0],
          [0, 219, 255],
          [175, 0, 255],
          [255, 233, 0],
          [0, 22, 255],
          [0, 255, 64],
          [0, 255, 141],
          [0, 255, 213],
          [255, 227, 0],
          [0, 255, 201],
          [223, 0, 255],
          [255, 167, 0],
          [199, 0, 255],
          [255, 0, 149],
          [0, 255, 177],
          [0, 58, 255],
          [255, 0, 197],
          [0, 255, 129],
          [44, 255, 0],
          [255, 24, 0],
          [0, 94, 255],
          [255, 215, 0],
          [86, 255, 0],
          [0, 255, 22],
          [253, 255, 0],
          [0, 255, 124],
          [0, 28, 255],
          [229, 255, 0],
          [211, 0, 255],
          [255, 191, 0],
          [20, 255, 0],
          [217, 255, 0],
          [0, 16, 255],
          [247, 255, 0],
          [56, 255, 0],
          [127, 255, 0],
          [2, 255, 0],
          [104, 0, 255],
          [38, 255, 0],
          [255, 114, 0],
          [0, 255, 171],
          [0, 231, 255],
          [68, 255, 0],
          [145, 255, 0],
          [157, 255, 0],
          [255, 0, 131],
          [151, 255, 0],
          [0, 100, 255],
          [0, 118, 255],
          [0, 255, 100],
          [62, 255, 0],
          [211, 255, 0],
          [0, 141, 255],
          [0, 255, 94],
          [44, 0, 255],
          [255, 0, 227],
          [0, 255, 70],
          [116, 0, 255],
          [0, 255, 58],
          [255, 0, 203],
          [0, 207, 255],
          [0, 255, 255],
          [0, 171, 255],
          [26, 255, 0],
          [110, 255, 0],
          [163, 0, 255],
          [0, 243, 255],
          [255, 0, 221],
          [0, 237, 255],
          [255, 0, 12],
          [255, 161, 0],
          [255, 48, 0],
          [193, 0, 255],
          [255, 108, 0],
          [0, 255, 82],
          [255, 36, 0],
          [0, 201, 255],
          [0, 255, 10],
          [0, 255, 118],
          [0, 255, 106],
          [255, 0, 36],
          [255, 0, 42],
          [255, 102, 0],
          [0, 153, 255],
          [0, 195, 255],
          [0, 255, 34],
          [20, 0, 255],
          [0, 4, 255],
          [199, 255, 0],
          [0, 255, 231],
          [0, 112, 255],
          [26, 0, 255],
          [32, 0, 255],
          [145, 0, 255],
          [255, 0, 60],
          [229, 0, 255],
          [241, 0, 255],
          [0, 106, 255],
          [98, 0, 255],
          [255, 0, 126],
          [205, 255, 0],
          [0, 159, 255],
          [0, 255, 189],
          [139, 255, 0],
          [8, 255, 0],
          [255, 0, 251],
          [255, 96, 0],
          [253, 0, 255],
          [255, 0, 96],
          [255, 0, 30],
          [14, 255, 0],
          [205, 0, 255],
          [50, 0, 255],
          [255, 0, 179],
          [255, 0, 84],
          [0, 52, 255],
          [0, 255, 40],
          [139, 0, 255],
          [187, 0, 255],
          [0, 255, 219],
          [0, 147, 255],
          [98, 255, 0],
          [2, 0, 255],
          [193, 255, 0],
          [255, 0, 120],
          [0, 40, 255],
          [255, 0, 137],
          [0, 255, 16],
          [187, 255, 0],
          [0, 255, 243],
          [0, 255, 28],
          [0, 46, 255],
          [0, 34, 255],
          [169, 0, 255],
          [255, 185, 0],
          [255, 221, 0],
          [255, 18, 0],
          [0, 70, 255],
          [255, 137, 0],
          [68, 0, 255],
          [255, 149, 0],
          [255, 0, 90],
          [241, 255, 0],
          [122, 255, 0],
          [0, 213, 255],
          [255, 131, 0],
          [80, 255, 0],
          [38, 0, 255],
          [255, 251, 0],
          [86, 0, 255],
          [247, 0, 255],
          [255, 90, 0],
          [255, 203, 0],
          [0, 64, 255],
          [0, 255, 249],
          [255, 0, 102],
          [0, 255, 195],
          [169, 255, 0],
          [255, 0, 114],
          [80, 0, 255],
          [92, 0, 255],
          [0, 255, 225],
          [0, 129, 255],
          [255, 126, 0],
          [56, 0, 255],
          [128, 0, 255],
          [32, 255, 0],
          [0, 255, 88],
          [255, 155, 0],
          [0, 165, 255],
          [0, 82, 255],
          [0, 88, 255],
          [255, 0, 108],
          [255, 0, 239],
          [255, 0, 245],
          [255, 0, 173],
          [8, 0, 255],
          [235, 0, 255],
          [0, 255, 153],
          [255, 78, 0],
          [255, 0, 167],
          [255, 0, 66],
          [255, 0, 54],
          [223, 255, 0],
          [12, 11, 11],
          [255, 0, 209],
          [255, 0, 72],
          [163, 255, 0],
          [0, 76, 255],
          [74, 0, 255],
          [74, 255, 0],
          [0, 10, 255],
          [255, 30, 0],
          [151, 0, 255],
          [255, 197, 0],
          [133, 255, 0],
          [0, 255, 46],
          [0, 255, 147],
          [255, 120, 0],
          [175, 255, 0],
          [255, 173, 0],
          [255, 0, 215],
          [181, 0, 255],
          [235, 255, 0],
          [104, 255, 0],
          [255, 54, 0],
          [0, 177, 255],
          [255, 209, 0],
          [0, 255, 135],
          [255, 0, 155],
          [0, 255, 165],
          [255, 0, 185],
          [255, 245, 0],
          [0, 255, 4],
          [255, 60, 0],
          [0, 255, 183],
          [255, 0, 161],
          [255, 239, 0],
          [110, 0, 255],
          [255, 0, 24],
          [255, 0, 48],
          [181, 255, 0],
          [0, 135, 255],
          [0, 124, 255],
          [255, 66, 0],
          [0, 255, 237],
          [255, 0, 6],
          [0, 189, 255],
          [255, 84, 0],
          [92, 255, 0],
          [0, 249, 255],
          [255, 6, 0],
          [14, 0, 255],
          [0, 255, 112],
          [50, 255, 0],
          [255, 143, 0],
          [255, 0, 78],
          [0, 255, 159],
          [133, 0, 255],
          [255, 0, 18],
          [0, 225, 255],
          [255, 0, 143],
          [0, 255, 76],
          [217, 0, 255],
          [157, 0, 255],
          [255, 0, 191],
          [122, 0, 255],
          [255, 179, 0],
          [62, 0, 255],
          [0, 255, 207],
          [0, 255, 52],
          [255, 0, 233],
          [255, 42, 0],
          [255, 72, 0],
          [0, 183, 255]]


def main_loop(dirpath, opath=None, startframe=0, endframe=None, tsdf=False, campose=None):
    bpy.data.objects.remove(bpy.data.objects['Cube'])
    bpy.data.objects.remove(bpy.data.objects['Light'])

    cam = bpy.data.objects['Camera'].data
    w, h = (640, 480)
    fx = fy = 525.
    cx = w/2 - .5
    cy = h/2 - .5

    calfile = os.path.join(dirpath, 'calibration.txt')
    if os.path.exists(calfile):
        with open(calfile, 'r') as f:
            line = f.readline()
            params = line.split(' ')
            fx, fy, cx, cy = map(float, params[:4])
            w, h = map(int, params[4:])

    cam.shift_x = -(cx / w - .5)
    cam.shift_y = (cy - .5 * h) / w

    cam.lens = fx / w * cam.sensor_width

    cam = bpy.data.objects['Camera']
    cam.location = mathutils.Vector((0,0,0))
    cam.rotation_euler = mathutils.Vector((math.pi,0,0))
    if campose is not None:
        loc, rot, _ = campose.decompose()
        cam.location = loc
        cam.rotation_euler.rotate(rot)

    light_data = bpy.data.lights.new(name='Light', type='AREA')
    light_data.energy = 500
    light_data.size = 10.24
    light = bpy.data.objects.new(name='Light', object_data=light_data)
    bpy.context.collection.objects.link(light)
    light.location = (0, -2.56, 2.56)
    light.rotation_euler = (1/2*math.pi, 0, 0)

    colormap = [mathutils.Color([c / 255 for c in rgb]) for rgb in colors]

    pixel_aspect = fy / fx
    scene = bpy.context.scene
    scene.render.resolution_x = w
    scene.render.resolution_y = h
    scene.render.pixel_aspect_x = 1.0
    scene.render.pixel_aspect_y = pixel_aspect

    camposes = {}
    cam_posefile = os.path.join(dirpath, 'poses-cam.txt')
    lastframe = -1
    with open(cam_posefile, 'r') as f:
        for line in f:
            parts = line.split(' ')
            t = int(parts[0])
            if t < startframe:
                continue

            loc = mathutils.Vector(map(float, parts[1:4]))
            rot = mathutils.Quaternion(map(float, [parts[i] for i in [7,4,5,6]]))
            camposes[t] = {'loc': loc, 'rot': rot}
            lastframe = t
            if t == endframe:
                break

    poses = {}
    obj = 1
    while True:
        posefile = os.path.join(dirpath, 'poses-' + str(obj) + '.txt')
        if not os.path.exists(posefile):
            break

        poses[obj] = {}
        with open(posefile) as f:
            for line in f:
                parts = line.split(' ')
                t = int(parts[0])
                if t < startframe:
                    continue

                loc = mathutils.Vector(map(float, parts[1:4]))
                rot = mathutils.Quaternion(map(float, [parts[i] for i in [7,4,5,6]]))
                poses[obj][t] = {'loc': loc, 'rot': rot}
        obj += 1

    frame = startframe
    last_keyframe = None
    meshes = {}
    while True:
        if campose is None:
            cam.location = camposes[frame]['loc']
            cam.rotation_euler = mathutils.Vector((math.pi,0,0))
            cam.rotation_euler.rotate(camposes[frame]['rot'])
        newFrame = frame if tsdf else (frame // 10) * 10
        if newFrame != last_keyframe:
            last_keyframe = newFrame
            for obj in bpy.data.objects:
                if obj.name != 'Light' and obj.name != 'Camera':
                    bpy.data.objects.remove(obj)
                meshes.clear()
            if tsdf:
                bg_meshfile = os.path.join(dirpath, 'frame_meshes', 'bg', '%04d.ply' % last_keyframe)
            else:
                bg_meshfile = os.path.join(dirpath, 'optim_meshes', 'bg', '%04d.ply' % last_keyframe)
            bpy.ops.import_mesh.ply(filepath=bg_meshfile)
            bg_mesh = bpy.context.selected_objects[0]
            bg_mesh.name = "BG"
            bg_mesh.location = mathutils.Vector((0, 0, 2.56))
            mat = bpy.data.materials.new(name="Mat bg")
            mat.use_nodes = True
            bg_mesh.data.materials.append(mat)
            #mat.diffuse_color = (.3, .3, .3, 1.0)
            #for f in bg_mesh.data.polygons:
                #f.use_smooth = True

            for obj in range(1, len(poses) + 1):
                if tsdf:
                    meshfile = os.path.join(dirpath, 'frame_meshes', str(obj), '%04d.ply' % last_keyframe)
                else:
                    meshfile = os.path.join(dirpath, 'optim_meshes', str(obj), '%04d.ply' % last_keyframe)

                if os.path.exists(meshfile):
                    bpy.ops.import_mesh.ply(filepath=meshfile)
                    meshes[obj] = bpy.context.selected_objects[-1]
                    meshes[obj].name = str(obj)
                    mat = bpy.data.materials.new(name="Mat " + str(obj))
                    mat.diffuse_color = (colormap[obj].b, colormap[obj].g, colormap[obj].r, 1)
                    mat.use_nodes = True
                    meshes[obj].data.materials.append(mat)
                    meshes[obj].rotation_mode = 'QUATERNION'
                    #for f in meshes[obj].data.polygons:
                        #f.use_smooth = True

        for obj, mesh in meshes.items():
            mesh.location = poses[obj][frame]['loc']
            mesh.rotation_quaternion = poses[obj][frame]['rot']

        bpy.context.scene.render.engine = 'CYCLES'
        bpy.context.scene.cycles.device = 'GPU'
        bpy.context.preferences.addons['cycles'].preferences.compute_device_type = 'CUDA'
        bpy.context.preferences.addons['cycles'].preferences.get_devices()
        bpy.context.scene.render.tile_x = 256
        bpy.context.scene.render.tile_y = 256
        bpy.context.scene.view_layers['View Layer'].cycles.use_denoising = True
        bpy.context.scene.render.filepath = os.path.join(opath, '%04d.png' % frame)
        bpy.ops.render.render(write_still=True)
        frame += 1
        if frame > lastframe:
            break


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Render 3D models created by "
                                                 "EM-Fusion or Co-Section")
    parser.add_argument(
        "--dir",
        "-d",
        dest="dir",
        required=True,
        help="Directory containing mesh and trajectory files"
    )
    parser.add_argument(
        "--opath",
        "-o",
        dest="opath",
        required=True,
        help="Output path. Should be a folder. Files will be written as opath/<framenum>.png"
    )
    parser.add_argument(
        "--frames",
        "-f",
        dest="frames",
        help="frames: either single number n or range n-m"
    )
    parser.add_argument(
        "--tsdf",
        "-t",
        dest="tsdf",
        action="store_true",
        default=False,
        help="Whether to use the TSDF mesh or the optimization output."
    )
    parser.add_argument(
        "--pose",
        "-p",
        dest="pose",
        help="The file storing the camera pose."
    )
    if '--' in sys.argv:
        argv = sys.argv[sys.argv.index('--') + 1:]
    #argv = utils.extract_args()
    args = parser.parse_args(argv)

    startframe = 0
    endframe = None

    if args.frames is not None:
        parts = args.frames.split("-")
        startframe = int(parts[0])
        if len(parts) > 1:
            endframe = int(parts[1])
        else:
            endframe = startframe

    campose = None
    if args.pose is not None:
        with open(args.pose, 'r') as f:
            m = [[float(n) for n in line.split()] for line in f]
            campose = mathutils.Matrix(m)

    main_loop(args.dir, args.opath, startframe, endframe, args.tsdf, campose)
